elastic/apm-server
# Drop `observer.id` and `observer.ephemeral_id`

## Motivation/summary

We don't use these fields in the UI anywhere, and for debugging we can rely on the hostname to distinguish one APM Server instance from another. The `observer.id` and `observer.ephemeral_id` fields suggest that we should care about specific instances (and think of them as long-lived), and their continued presence contradicts our intentions for a more modern, cloud native, approach, where APM Server instances may come and go.

For now we need to continue persisting a UUID to disk for stack monitoring, as this is used to identify instances of APM Server. In the future we expect to revisit how stack monitoring works, and at that point we should redesign the documents to not enshrine a "pets" mentality. We may still want to identify individual instances (e.g. for calculating time-series IDs), but statefulness should not be required.

For tail-based sampling we now generate a UUID (for identifying its documents) at process start. The field has been renamed from `observer.id` to `agent.ephemeral_id`, the latter of which is an ECS field (`observer.id` is not).

We no longer pass `beat.Info` from `beatcmd` into package `beater`. Instead, `beater` hard-codes `observer.type: apm-server`, uses package `version` for `observer.version`, and calls `os.Hostname()` for `observer.hostname`. We need to recreate `beat.Info` for the libbeat outputs. If at some point we drop support for libbeat outputs, then we would have no more dependencies on `beat.Info` outside of package `beatcmd`.

## Checklist


## How to test these changes

1. Ensure tail-based sampling still works (with multiple servers)
2. Ensure stack monitoring still works (with multiple servers)

## Related issues

Closes https://github.com/elastic/apm-server/issues/3874
